### PR TITLE
Add Olmo3 implementation

### PR DESCRIPTION
**Pull request by 2015aroras**
This PR adds the upcoming Olmo 3. The main architectural differences from Olmo 2 are:

- Sliding window attention is used for 3 out of 4 layers. RoPE scaling is not applied to sliding window attention layers.

Since the architecture is very similar to Olmo 2, this PR opts to merge Olmo 3 changes into the Olmo 2 implementation (similar to https://github.com/vllm-project/vllm/pull/24534). I can create a separate Olmo 3 implementation instead if preferred.

### CODE REVIEW

**Comment by 2015aroras** [comment]
I used the model conversion example for testing. I got the following results when using bf16 on [shanearora/2025-sep-a-base-model](https://huggingface.co/shanearora/2025-sep-a-base-model), modified to have yarn rope scaling enabled.

```
📈 METRICS
==============================
MSE (Mean Squared Error):     1.592396e-02
Reference Variance:           6.831117e+00
NMSE:                         2.331092e-03
Max Absolute Error:           0.438750
Mean Absolute Error:          0.116665
NMSE (dB):                    -26.32 dB

🎯 INTERPRETATION
==============================
👍 Good match

📋 GUIDANCE
==============================
👍 GOOD: Your GGML conversion is working well.
   Small differences are likely due to precision/quantization.

📚 NMSE BENCHMARKS
==============================
✅ RESULT: PASS (NMSE = 2.33e-03)
```

Also, below is the [allenai/OLMo-2-0425-1B](https://huggingface.co/allenai/OLMo-2-0425-1B) with fp32.

```
📈 METRICS
==============================
MSE (Mean Squared Error):     1.594746e-03
Reference Variance:           9.219801e+00
NMSE:                         1.729697e-04
Max Absolute Error:           0.168732
Mean Absolute Error:          0.033951
NMSE (dB):                    -37.62 dB

🎯 INTERPRETATION
==============================
👍 Very good match

📋 GUIDANCE
==============================
✅ EXCELLENT: Your GGML conversion is working very well!
   The differences are negligible for practical use.

📚 NMSE BENCHMARKS
==============================
✅ RESULT: PASS (NMSE = 1.73e-04)
```

File: src/llama-model.cpp [hunk 1]
```diff
@@ -12197,25 +12213,45 @@ struct llm_build_olmo2 : public llm_graph_context {
...
+                        0.0, 1.0, beta_fast, beta_slow
+                        );
+
+                    Kcur = ggml_rope_ext(
+                        ctx0, Kcur, inp_pos, nullptr,
+                        n_rot, rope_type, n_ctx_orig, freq_base, 1.0,
+                        0.0, 1.0, beta_fast, beta_slow
+                        );
+                }
+                else {
```
Comment (CISC) [inline src/llama-model.cpp]:
```suggestion
                if (!is_swa) {
```

File: src/llama-model.cpp [hunk 2]
```diff
@@ -12197,25 +12213,45 @@ struct llm_build_olmo2 : public llm_graph_context {
...
 
                 cb(Qcur, "Qcur", il);
                 cb(Kcur, "Kcur", il);
                 cb(Vcur, "Vcur", il);
 
                 cur = build_attn(inp_attn,
-                        model.layers[il].wo, NULL,
-                        Qcur, Kcur, Vcur, nullptr, nullptr, nullptr, 1.0f/sqrtf(float(n_embd_head)), il);
+                    model.layers[il].wo, NULL,
+                    Qcur, Kcur, Vcur, nullptr, nullptr, nullptr, 1.0f/sqrtf(float(n_embd_head)), il);
```
Comment (CISC) [inline src/llama-model.cpp]:
```suggestion
                        model.layers[il].wo, NULL,
                        Qcur, Kcur, Vcur, nullptr, nullptr, nullptr, 1.0f/sqrtf(float(n_embd_head)), il);
```

**Review by CISC — APPROVED** [review]

File: src/llama-model.cpp (hunk 1 above)
Comment (2015aroras) [reply src/llama-model.cpp]: This if block is needed. For SWA layers Olmo2 uses standard rope. Removing that if block would remove rope on SWA layers.

Comment (2015aroras) [reply src/llama-model.cpp]: https://github.com/ggml-org/llama.cpp/pull/16015/commits/6997fad8112f78d8ca2cb812d99d8028823857d6 Clarified comment slightly

File: src/llama-model.cpp (hunk 2 above)
Comment (2015aroras) [reply src/llama-model.cpp]: Applied this, though frankly the indentation looks worse?

**Comment by pwilkin** [comment]
@2015aroras What tool are you using to compare the conversion?

**Comment by 2015aroras** [comment]
@pwilkin I am using the [model conversion tools inside this repo](https://github.com/ggml-org/llama.cpp/blob/master/examples/model-conversion/README.md). These have been created to help make sure HF to llama.cpp conversion is accurate. The logs above are from the [Model logits verfication](https://github.com/ggml-org/llama.cpp/blob/master/examples/model-conversion/README.md#model-logits-verfication) step.

**Comment by pwilkin** [comment]
Ah, that's nice, I haven't used that specific one yet :)

**Comment by 2015aroras** [comment]
All the check failures seem to be unrelated to this change. Before merging master again, an ios check was failing instead. So imo ready to merge.

File: src/llama-model.cpp [hunk 3]
```diff
@@ -12281,17 +12297,37 @@ struct llm_build_olmo2 : public llm_graph_context {
...
+                        0.0, 1.0, beta_fast, beta_slow
+                        );
+
+                    Kcur = ggml_rope_ext(
+                        ctx0, Kcur, inp_pos, nullptr,
+                        n_rot, rope_type, n_ctx_orig, freq_base, 1.0,
+                        0.0, 1.0, beta_fast, beta_slow
+                        );
+                }
+                else {
```
Comment (CISC) [inline src/llama-model.cpp]:
```suggestion
                } else {
```

**Comment by CISC** [comment]
> All the check failures seem to be unrelated to this change. Before merging master again, an ios check was failing instead. So imo ready to merge.

Yes, sorry for the delay, just a minor cosmetic change and we'll merge. :)
